dm/worker/meta.go

```diff
@@ -74,95 +77,240 @@ func (m *Meta) adjust() error {
+	log, err := meta.log.GetTaskLog(meta.db, opLogID)
+	return log, errors.Trace(err)
+}
+
+// PeekLog returns first need to be handled task log
+func (meta *Metadata) PeekLog() (log *pb.TaskLog) {
+	meta.RLock()
+	if len(meta.logs) > 0 {
+		log = CloneTaskLog(meta.logs[0])
+	}
+	meta.RUnlock()
+
+	return
+}
+
+// AppendOperation appends operation into task log
+func (meta *Metadata) AppendOperation(subTask *pb.TaskMeta) (int64, error) {
+	meta.Lock()
+	defer meta.Unlock()
+
+	opLog := &pb.TaskLog{
+		Task: CloneTaskMeta(subTask),
+	}
+
+	if err := meta.log.Append(meta.db, opLog); err != nil {
+		return 0, errors.Trace(err)
+	}
+
+	meta.logs = append(meta.logs, opLog)
+	return opLog.Id, nil
+}
+
+// MarkOperation marks operation result
+func (meta *Metadata) MarkOperation(log *pb.TaskLog) error {
+	meta.Lock()
+	defer meta.Unlock()
+
+	if len(meta.logs) == 0 {
+		return errors.NotFoundf("any task operation log")
+	}
+
+	if meta.logs[0].Id != log.Id {
+		return errors.Errorf("please handle task operation order by log ID, the log need to be handled is %+v, not %+v", meta.logs[0], log)
+	}
+
+	txn, err := meta.db.OpenTransaction()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = meta.log.MarkAndForwardLog(txn, log)
 	if err != nil {
+		txn.Discard()
 		return errors.Trace(err)
 	}
 
-	if err := ioutil.WriteFile(metaDB.path, []byte(serialized), 0644); err != nil {
+	if log.Success {
+		if log.Task.Op == pb.TaskOp_Stop {
+			err = DeleteTaskMeta(txn, log.Task.Name)
+		} else {
+			err = SetTaskMeta(txn, log.Task)
+		}
+		if err != nil {
+			txn.Discard()
+			return errors.Trace(err)
+		}
+	}
+
+	err = txn.Commit()
+	if err != nil {
 		return errors.Trace(err)
 	}
+
+	if log.Success {
+		if log.Task.Op == pb.TaskOp_Stop {
+			delete(meta.tasks, log.Task.Name)
+		} else {
+			meta.tasks[log.Task.Name] = log.Task
+		}
+	}
+	meta.logs = meta.logs[1:]
 	return nil
 }
 
-// Get returns `Meta` object
-func (metaDB *FileMetaDB) Get() *Meta {
-	metaDB.lock.Lock()
-	defer metaDB.lock.Unlock()
-
-	meta := &Meta{
-		SubTasks: make(map[string]*config.SubTaskConfig),
+func (meta *Metadata) loadFromDB() (err error) {
+	meta.logs, err = meta.log.Initial(meta.db)
+	if err != nil {
+		return errors.Trace(err)
 	}
 
-	for name, task := range metaDB.meta.SubTasks {
-		meta.SubTasks[name] = task
+	meta.tasks, err = LoadTaskMetas(meta.db)
+	if err != nil {
+		return errors.Trace(err)
 	}
 
-	return meta
+	return nil
 }
 
-// Set sets subtask in Meta
-func (metaDB *FileMetaDB) Set(subTask *config.SubTaskConfig) error {
-	metaDB.lock.Lock()
-	defer metaDB.lock.Unlock()
+// to be compatible with the old fashion meta
+func (meta *Metadata) tryToRecoverMetaFromOldFashion(path string) error {
+	fd, err := os.Open(path)
+	// old metadata file exists, recover metadata from it
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return errors.Trace(err)
+		}
+		return nil
+	}
+	fd.Close()
 
-	metaDB.meta.SubTasks[subTask.Name] = subTask
+	data, err := ioutil.ReadFile(path)
```

Comment:
Can L270~278 be replaced by handing this `err`? then only need to open the file once.